Improve sitemap

cdedreuille opened the pull request:
- This PR moves the creation of sitemaps using Next.js instead of `next-sitemap`.
- It also improves the way trailing slashes are added at the end
- We removed drafts pages from docs
- We removed directories on level 2 from docs
- New sitemap architecture to make it more understandable.

kylegach commented on apps/frontpage/app/sitemap-all/sitemap.xml:
```diff
@@ -0,0 +1,11 @@
+<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
```
I don't understand why this isn't `app/sitemap-all.xml`?

If you did that, everything would be under the same directory (the site root). Thus, there'd be no need to generate duplicate sitemaps for addons and recipes, and the docs one could live at `app/docs/sitemap-all.xml`.

cdedreuille replied:
You're right. The trick to make this happen is to put `sitemap-all.tsx` in the public folder since `sitemap.xl` is a special naming. I just can't put it next to `sitemap.xml`.

cdedreuille replied:
This is now fixed.

kylegach commented on apps/frontpage/app/blog/sitemap.ts:
```diff
@@ -0,0 +1,19 @@
+      const url = site.replace(/\/$/, '');
+      return { url };
+    });
+  } catch (error) {
```
If there's an error we need to fail the build. We cannot ship an empty sitemap without SEO consequences.

kylegach replied:
Applies multiple places

kylegach commented on apps/frontpage/app/docs/sitemap.ts:
```diff
@@ -0,0 +1,40 @@
+  flattenTree(tree);
+
+  const filteredTree = flattenedTree
+    .filter((node) => node.type !== 'directory')
```
Does this also filter out pages like https://storybook.js.org/docs/writing-stories ? Because those _should_ be indexed.

It's only the second-level and below, e.g. https://storybook.js.org/docs/writing-stories/mocking-data-and-modules, directory nodes that we won't to filter out.

cdedreuille replied:
You're right. I was filtering too much. I added the level in the loop to make sure we can filter only level 2

cdedreuille replied:
Now we are filtering if this is a director + if this is only level 2.

kylegach commented on apps/frontpage/app/docs/sitemap.ts:
```diff
@@ -0,0 +1,40 @@
+    .filter((node) => node.draft !== true);
+
+  const docs = filteredTree.map((node) => ({
+    url: `https://storybook.js.org${node.slug}`.replace(
+      `${latestVersion.id}/`,
+      '',
+    ),
```
Not blocking

TODO: Refactor this to ensure it only ever replaces versions at the start of the route. If a page slug ever contains a version id, e.g. `migrate-from-8.0`, this will break.

cdedreuille replied:
The real refactor that need to happen is to make sure the slug reflect the real scenario so we don't have to do any modifications here.

cdedreuille replied:
This should not be much better.

kylegach commented on apps/frontpage/app/sitemap-all/docs/sitemap.ts:
```diff
@@ -0,0 +1,52 @@
+    const url = `https://storybook.js.org${node.slug}`
+      .replace(`${latestVersion.id}/`, '')
+      .replace(docsVersions[1].id, docsVersions[1].inSlug || '')
+      .replace(docsVersions[2].id, docsVersions[2].inSlug || '')
```
We don't currently have one, but we _can_ have a pre-release version, which would mean there's a `docsVersions[3]`.

It makes me nervous to have hard-coded array lengths like this instead of mapping over an array that could be any length.

cdedreuille replied:
The real refactor that need to happen is to make sure the slug reflect the real scenario so we don't have to do any modifications here.

cdedreuille replied:
You're right. I forgot that we had a `getUrl` function that we made for the sidebar. I made it globally available so we can use it here as well.

kylegach commented on apps/frontpage/app/sitemap-all/docs/sitemap.ts:
```diff
@@ -0,0 +1,52 @@
+      .replace(`${latestVersion.id}/`, '')
+      .replace(docsVersions[1].id, docsVersions[1].inSlug || '')
+      .replace(docsVersions[2].id, docsVersions[2].inSlug || '')
+      .replace(/index$/, '') // This line removes 'index' at the end of the URL
+      .replace(/\/$/, '');
```
Why is this not also done for `app/docs/sitemap.ts`?

cdedreuille replied:
Because all urls in docs are coming from Next.js and don;t have a trailing slash. I can add it for safety but that's just not going to do anything.

kylegach reviewed (changes requested):
This architecture is undoubtedly better!

I left some comments that must be addressed before this can be merged.

cdedreuille commented:
@kylegach Thanks a lot for your feedbacks. I addressed all the points you shared above. The sitemap feels way more robust now. I'm going to merge it to make some tests today but if you have any concerns please let me know.
